File: go/080-meetup/meetup.go
package meetup

import (
	"time"
)

type WeekSchedule int

const (
	First WeekSchedule = iota
	Second
	Third
	Fourth
	Teenth
	Last
)

func Day(wSched WeekSchedule, wDay time.Weekday, month time.Month, year int) int {
	day := time.Date(year, month, 1, 0, 0, 0, 0, time.UTC)

	if wSched == Teenth {
		for i := 0; i < 7; i++ {
			finalDay := day.AddDate(0, 0, 12+i)
			if finalDay.Weekday() == wDay {
				return finalDay.Day()
			}
		}
	}

	var foundDay []int
	for i := 0; i < 31; i++ {
		finalDay := day.AddDate(0, 0, i)
		if finalDay.Weekday() == wDay && finalDay.Month() == day.Month() {
			foundDay = append(foundDay, finalDay.Day())
		}
	}

	if wSched < 4 {
		return foundDay[wSched]
	}

	return foundDay[len(foundDay)-1]
}

File: ruby/102-meetup/meetup.rb
require 'date'

class Meetup
  # @param[Numeric]month
  # @param[Numeric]year
  def initialize(month, year)
    @month = month
    @year = year
  end

  def day(weekday, schedule)
    days = find_days_of_week(weekday)
    
    case schedule
    when :teenth
      days.find { |day| (13..19).include?(day.mday) }
    when :first
      days.first
    when :second
      days[1]
    when :third
      days[2]
    when :fourth
      days[3]
    when :last
      days.last
    end
  end

  private

  def find_days_of_week(weekday)
    day_number = day_to_number(weekday)
    
    # Create a range for all days in the month
    first_day = Date.new(@year, @month, 1)
    last_day = Date.new(@year, @month, -1)
    
    # Find all days that match the weekday
    (first_day..last_day).select { |date| date.wday == day_number }
  end

  def day_to_number(day)
    {
      monday: 1,
      tuesday: 2,
      wednesday: 3,
      thursday: 4,
      friday: 5,
      saturday: 6,
      sunday: 0
    }[day]
  end
end